Week3/HW03.hs

```diff
@@ -0,0 +1,169 @@
+module HW03 where
+
+-- https://www.cis.upenn.edu/~cis194/spring15/hw/03-ADTs.pdf
+
+import Data.Bool
+
+data Expression =
+    Var String                   -- Variable
+  | Val Int                      -- Integer literal
+  | Op Expression Bop Expression -- Operation
+  deriving (Show, Eq)
+
+-- Binary (2-input) operators
+data Bop = 
+    Plus     
+  | Minus    
+  | Times    
+  | Divide   
+  | Gt
+  | Ge       
+  | Lt  
+  | Le
+  | Eql
+  deriving (Show, Eq)
+
+data Statement =
+    Assign   String     Expression
+  | Incr     String
+  | If       Expression Statement  Statement
+  | While    Expression Statement       
+  | For      Statement  Expression Statement Statement
+  | Sequence Statement  Statement        
+  | Skip
+  deriving (Show, Eq)
+
+type State = String -> Int 
+
+-- Exercise 1 -----------------------------------------
+
+extend :: State -> String -> Int -> State
+extend initState name val a             = bool (initState a) val (a == name)
+
+empty :: State
+empty _                                 = 0
+
+-- Exercise 2 -----------------------------------------
+
+toInt :: Bool -> Int
+toInt b = if b then 1 else 0 
+
+evalE :: State -> Expression -> Int
+evalE _ (Val a) = a 
+evalE initState (Op exp1 bop exp2) = 
+  let a = evalE initState exp1 
+      b = evalE initState exp2
+  in case bop 
+    of  Plus            -> a + b
+        Minus           -> a - b
+        Times           -> a * b
+        Divide          -> a `div` b
+        Gt              -> toInt $ a > b
+        Ge              -> toInt $ a >= b
+        Lt              -> toInt $ a < b
+        Le              -> toInt $ a <= b 
+        Eql             -> toInt $ a == b
+evalE _ _ = 0
```

Comment:
You're missing a case for `(Var x)`. Did you notice, that you're never actually looking up any values in state that you're passing around?

Reply:
Right, got it.